Preview Sites: Delete all preview sites using `/delete/all` API endpoint

gavande1 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

- Fixes STU-851

## Proposed Changes

- Add a new mutation endpoint for deleting all preview sites for the user 
- Use the API endpoint to delete preview sites in the preference modal

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Apply the patch 194679-ghe-Automattic/wpcom on your sandbox.
2. Run Studio using this branch.
3. Start the application with:

   ```bash
   npm start
   ```
4. Open Developer Tools and keep the Network tab open.
5. Navigate to the Preferences tab in Studio and log in (if not already).
6. Go to the Usage tab and click Delete all preview sites.
7. Verify that a request to

   ```
   /jurassic-ninja/delete/all/delete/all
   ```

   was sent and returned a successful response
8. Confirm that the usage numbers are reset.
9. Close and reopen the Usage tab.
10. Ensure the updated usage numbers are still reflected.
11. Open the following file on your system:

    ```
    ~/Library/Application Support/Studio/appdata-v1.json
    ```
12. Verify that the snapshots array is cleared and shows as `[]`.


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

gavande1 committed:
Add offline check for react mutation

gavande1 committed:
Add new mutation for deleting all preview sites

gavande1 committed:
Use newly created mutation for deleting user's all previews sites

gavande1 committed:
Refactoring error handling

gavande1 committed:
Add snapshots to allowed list

gavande1 committed:
Clear app data on successful request

gavande1 committed:
Do not clear current state

gavande1 committed:
Fix linting issues

wojtekn commented:
I was wondering if it will be clear to the user that it removes all preview sites on the account, and not only those created from the given Studio instance, but it seems that the dialog already makes it clear:

> All preview sites that exist for your WordPress.com account, along with all posts, pages, comments, and media, will be lost.

katinthehatsite commented on src/stores/utils/with-offline-check.ts:
```diff
@@ -13,3 +13,23 @@ export function withOfflineCheck< TResult, TArg, TBaseQuery extends BaseQueryFn
 	};
 }
+
+export function withOfflineCheckMutation< TResult, TArg, TBaseQuery extends BaseQueryFn >(
```
We already disable the button to delete preview sites when the user isOffline, so I am wondering if we need this? 

<img width="636" height="240" alt="Image" src="https://github.com/user-attachments/assets/41e71626-510d-4db4-9b22-9e335f2ebddd" />

gavande1 replied:
That’s a great point. I agree it’s not needed right now, but I added it just in case we need the offline check elsewhere while using another API. No strong opinion here, though I would slightly prefer keeping it since `withOfflineCheckMutation` could be useful later.

katinthehatsite replied:
Okay, we can keep it then 👍 If it is not something that we use often, we can clean this up later.

gavande1 commented:
> I was wondering if it will be clear to the user that it removes all preview sites on the account, and not only those created from the given Studio instance, but it seems that the dialog already makes it clear:

This is what I thought as well. That's why I didn't alter the message.

ivan-ottinger reviewed (commented):
The changes look good and work correctly. The new endpoint is being used:

<img width="1750" height="258" alt="Screen Shot on 2025-10-29 at 11:02:48" src="https://github.com/user-attachments/assets/d3e58fb9-afd5-4f0f-b031-aa4e26180874" />

and the logic clears up the `snapshots` in the `appdata-v1.json` as well:
<img width="602" height="188" alt="Screen Shot on 2025-10-29 at 11:04:47" src="https://github.com/user-attachments/assets/bf768ca0-4201-4617-b56c-1402955d783b" />

One thing I noticed is that the `Delete all preview sites` button does not get disabled when there are no preview sites left:

<img width="1068" height="796" alt="Screen Shot on 2025-10-29 at 11:02:01" src="https://github.com/user-attachments/assets/e6a87fc1-5f3e-4df4-93b7-d9c750a3a6c2" />

gavande1 commented:
> One thing I noticed is that the Delete all preview sites button does not get disabled when there are no preview sites left:

@ivan-ottinger, Great question. This is the intended behavior for cases where two separate instances of Studio are out of sync. While disabling the button might seem reasonable, there are several scenarios where keeping it enabled is actually helpful.

ivan-ottinger commented:
> @ivan-ottinger, Great question. This is the intended behavior for cases where two separate instances of Studio are out of sync. While disabling the button might seem reasonable, there are several scenarios where keeping it enabled is actually helpful.

I see. Are you referring to a case where the `snapshots`  are still recorded on some other app instance? Is the intention to clear that leftover `snapshosts` array with the button?

wojtekn commented:
> I see. Are you referring to a case where the snapshots are still recorded on some other app instance? Is the intention to clear that leftover snapshosts array with the button?

This is the goal, with the new implementation, the button should always be available to let the user clean preview sites from their account.

ivan-ottinger commented:
> This is the goal, with the new implementation, the button should always be available to let the user clean preview sites from their account.

Makes sense. 👍🏼

katinthehatsite commented:
> This is the goal, with the new implementation, the button should always be available to let the user clean preview sites from their account.

I am finding this a bit of a strange UI but I guess there are not many ways around it.

katinthehatsite reviewed (approved):
👍

ivan-ottinger commented:
> I am finding this a bit of a strange UI but I guess there are not many ways around it.

Yeah. likewise. I thought about it yesterday and maybe I am missing something, but:

Cannot we do the following?
- disable the `Delete all preview sites` button when there are no preview site linked to the user's account
- let the `Clear` button on the Previews tab remove those already-deleted sites from `appdata-v1.json` on another Studio instance

The `Delete all preview sites` button would:
- trigger the endpoint to clear all preview sites on the server
- clear preview sites from `appdata-v1.json`  on the local Studio instance

The `Clear` button on the Previews tab would only clear the leftover preview sites from the `appdata-v1.json` .

gavande1 commented:
> Cannot we do the following?
> - disable the Delete all preview sites button when there are no preview site linked to the user's account
> - let the Clear button on the Previews tab remove those already-deleted sites from appdata-v1.json on another Studio instance

Since we are already [querying usage](https://github.com/Automattic/studio/blob/trunk/src/modules/user-settings/components/snapshot-info.tsx#L26) on the component. I think it makes sense to do a quick check on the user's account. I will update the code to disable the button when there are no preview sites on the user's account. cc @wojtekn 

The remaining suggested behaviours have already been implemented in #1950.

gavande1 committed:
Disable delete all button when user has no snapshots

gavande1 commented:
@ivan-ottinger Thanks for the feedback so far. I have addressed your concerns in [e45a445](https://github.com/Automattic/studio/pull/1950/commits/e45a445ae217073ad5eeceb5d20b533613366456). Could you review it again?

gavande1 committed:
Fix linting issue

ivan-ottinger commented:
> @ivan-ottinger Thanks for the feedback so far. I have addressed your concerns in [e45a445](https://github.com/Automattic/studio/pull/1950/commits/e45a445ae217073ad5eeceb5d20b533613366456). Could you review it again?

Nice! Thank you for the updates. The changes look good and I can see that if there are no remote preview sites, the button gets disabled:

<img width="1084" height="754" alt="Screen Shot on 2025-10-30 at 12:55:58" src="https://github.com/user-attachments/assets/5fde371f-1d0d-4602-accf-14e2f13c62c1" />
